### ISSUE TITLE
atomWithQuery, how to handle invalid input from atom getter?

### ISSUE THREAD

**Issue by asutula** [jotaijs/jotai-trpc#12]
I'm new to Jotai, so this may be an issue with me not understanding more general patterns in Jotai, but let me explain.

I'm trying to create a trpc atom that gets its input value from another atom that sometimes holds a valid input value but holds `null` at other times. For example, it's a user id atom that holds a user id number when the user is logged in, but is null when the user logs out.

```
const userIdAtom = atom<number | null> = atom(null);

const userTeamsAtom = trpcJotai.teams.teamsForUser.atomWithQuery((get) => {
  const userId = get(userIdAtom);
  if (userId) {
    return { userId };
  } else {
    ???????
  }
});
```

How can I prevent the query atom from running when there is no valid input provided by `userIdAtom`? Is there some Jotai usage pattern to deal with situations like this?

Thank you.

**Comment by asutula** [comment]
So far I'm coming up with something like:
```
const userTeamsAtom = atom((get) => {
  const userId = get(userIdAtom);
  if (userId) {
    return get(
      trpcJotai.teams.teamsForUser.atomWithQuery({ userId })
    );
  }
  return new Promise<
    {
      label: string;
      teams: {
        id: string;
        name: string;
      }[];
    }[]
  >((resolve) => resolve([]));
});
```

**Comment by asutula** [comment]
Oh it seems I don't need to return the promise with empty results and the resulting atom value is `Team | undefined`. Nice.

**Comment by dai-shi** [comment]
Yeah, wrapping with another atom would be only the solution for now.

As trpc/client doesn't support it, it might make sense that jotai-trpc supports it such as `disabled` option.

**Comment by asutula** [comment]
Ok thanks @dai-shi. I'll close this for now since it sounds like my solution is the right way for now.

**Comment by dai-shi** [comment]
Cool.

Well, while you are here, let's see if `disabled` option works. I'll open a PR soon.

**Comment by asutula** [comment]
Ok sure. I'll give it a try as soon as i hear back from you.

**Comment by dai-shi** [comment]
@asutula Please check #14.

You can install the codesandbox build: https://ci.codesandbox.io/status/jotaijs/jotai-trpc/pr/14

**Comment by asutula** [comment]
Hey @dai-shi, it's not clear to me how to actually use that new option. My first thought was it would be available as a property in the options object optionally passed to `atomWithQuery`, but I still just see `context` and `signal` there. If I were more familiar with the Jotai codebase and Typescript generics, I could probably get further in figuring this out, but at this point, it's just faster to ask you for some guidance. Thanks.

**Comment by kdurek** [comment]
Yeah i got the same feeling, I wanted to do something on Error and can't make it work

**Comment by dai-shi** [comment]
> Hey @dai-shi, it's not clear to me how to actually use that new option. My first thought was it would be available as a property in the options object optionally passed to `atomWithQuery`, but I still just see `context` and `signal` there. If I were more familiar with the Jotai codebase and Typescript generics, I could probably get further in figuring this out, but at this point, it's just faster to ask you for some guidance. Thanks.

Oops, I forgot to change types...

It should be something like this:
```ts
const userIdAtom = atom<number | null> = atom(null);

const userTeamsAtom = trpcJotai.teams.teamsForUser.atomWithQuery(
  (get) => {
    const userId = get(userIdAtom) || '';
    return { userId };
  },
  (get) => {
    const userId = get(userIdAtom);
    if (userId === null) {
      return { disabled: true, defaultOption: /* hmm, what should be here? */ };
    }
    return {};
  }
);
```

On second thought, would it be easier to make null query and null output? Something like:
```ts
const userIdAtom = atom<number | null> = atom(null);

const userTeamsAtom = trpcJotai.teams.teamsForUser.atomWithQuery(
  (get) => {
    const userId = get(userIdAtom);
    if (userId === null) {
      return null; // means "disabled"
    }
    return { userId };
  }
);
```

**Comment by dai-shi** [comment]
> Yeah i got the same feeling, I wanted to do something on Error and can't make it work Is it possible to make it pass full react query options?

With Jotai atoms, errors are just thrown. You can either use ErrorBoundary or use `loadable` util. jotai-trpc doesn't use react query at all. (It doesn't even depend on React.)

**Comment by kdurek** [comment]
> > Yeah i got the same feeling, I wanted to do something on Error and can't make it work Is it possible to make it pass full react query options?
> 
> 
> 
> With Jotai atoms, errors are just thrown. You can either use ErrorBoundary or use `loadable` util. jotai-trpc doesn't use react query at all. (It doesn't even depend on React.)

Oh thank you for information, I didn't realize that 😁

**Comment by asutula** [comment]
> On second thought, would it be easier to make null query and null output?
While I think both solutions require the use of "magic" value that needs to understood by the developer, this idea about retuning `null` to disable definitely seems more simple and easy to understand. In that case, does the atom still need to return some default value or is there a way to just swallow the `null` input, disable, and not notify anyone using the atom? (This is a general question I've had with atoms... Is there a way in one atom to prevent further processing of the composed chains of atoms?)

**Comment by dai-shi** [comment]
With the second approach, the output will be `null` if disabled, so you need to deal with it.

> Is there a way in one atom to prevent further processing of the composed chains of atoms?

Only the way is if you return the same (`===`) value as the previous one, the propagation will be stopped.

**Comment by dai-shi** [comment]
Hmm, the input can really be `null` even if it's not disabled. So, we can't take the second approach. Let me continue with the first approach.

**Comment by dai-shi** [comment]
@asutula #14 is updated.

Please try the codesandbox build: https://ci.codesandbox.io/status/jotaijs/jotai-trpc/pr/14

**Comment by asutula** [comment]
Yea, ok I can see how to use it now. My code could look like this:

```
const authAtom = atom<string | null>(null);

const userTeamsAtom = trpcJotai.teams.teamsForUser.atomWithQuery(
  (get) => {
    const auth = get(authAtom);
    return { userId: auth || "" };
  },
  (get) => {
    const auth = get(authAtom);
    return { disabled: auth === null, disabledOutput: [] };
  }
);
```

I really appreciate you providing possible solutions to this problem. My honest feeling about this particular solution, though, is that it doesn't feel right. It feels like I'm repeating myself in the the input getter and options getter, and being forced to return some meaningless `''` value from the input getter doesn't feel right. That being said, I am clearly not an expert on Jotai, so if you believe this is the best possible solution, I'm happy with that.

I have a two more thoughts:

One API I thought about that would feel better to me would look more like this:
```
const authAtom = atom<string | null>(null);

const userTeamsAtom = trpcJotai.teams.teamsForUser.atomWithQuery(
  (get) => {
    const auth = get(authAtom);
    if (auth !== null) {
        return { input: { userId: auth } }
    } else {
        return { disabled: true, disabledOutput: [] }
    }
);
```
So basically the input getter would return a type that looks like:
```
type InputGetter<T> = {
    input?: T;
    disabled?: boolean;
    disabledOutput?: unknown;
}
```
I guess that disabled information could be modeled in a couple different ways, but you get the idea. The goal here is there is no repetition in two different getters and no arbitrary value being returned from the input getter when there is no valid input to get. But of course, I have no idea if this is possible with the architecture of Jotai.

The last thing to bring up is: I just noticed the input getter can't be `async`. That is a problem for me because my `authAtom` is actually `async`. Is it possible for input getter be `async`?

Thanks again for all your work here!

**Comment by dai-shi** [comment]
> I guess that disabled information could be modeled in a couple different ways, but you get the idea.

Yeah, I don't prefer changing the input structure. I think I'd use a symbol.

> Is it possible for input getter be async?

Yeah, it's actually easy.

**Comment by dai-shi** [comment]
#14 is updated.

Not sure how TS works, but I imagine the usage is:
```ts
const authAtom = atom<string | null>(null);

const userTeamsAtom = trpcJotai.teams.teamsForUser.atomWithQuery(
  (get) => {
    const auth = get(authAtom);
    if (auth !== null) {
        return { userId: auth };
    } else {
        return DISABLED;
    }
);
```

**Comment by asutula** [comment]
Yea this is looking great, feels good. It's actually looking a lot like your original alternate suggestion when you said
> On second thought, would it be easier to make null query and null output?

And upon testing out this latest version, I actually think that returning `undefined` (or maybe `null`, not sure which makes more sense) would be better. That way we don't have to import this `DISABLED` symbol in the store and any component that uses the atom, and we don't have to deal with the possible `DISABLED` value in the component using the atom (because the type when using the atom in a component is `T | DISABLED`). It would be easier to just deal with `T | undefined` so we can use optional chaining or conditionals.

The async support is great. This is feeling good. One more commit and I think we got it.

**Comment by asutula** [comment]
It's amazing that the types work out in a way where, on the consuming end of the atom in a component, you only have to deal with the disabled value type (`T | DISABLED` or `T | undefined/null`) if your input getter actually can return `DISABLED/undefined/null`. Other wise the type is just `T`. So good.

**Comment by dai-shi** [comment]
Did you try if it really works as expected? (Because I didn't test it...)
https://ci.codesandbox.io/status/jotaijs/jotai-trpc/pr/14

> I actually think that returning undefined (or maybe null, not sure which makes more sense) would be better. 

`null` was my original idea, but technically, trpc procedure can return `null` as I understand, and it conflicts. We can make it option to be both safe and easy.

**Comment by asutula** [comment]
Yes, just tested it and it seems to work as expected. In my component, the code looks like this in order to deal with `DISABLED`:
```
const [teams] = useAtom(userTeamsAtom);
const finalTeams = teams === DISABLED ? undefined : teams;
```
Then I use `finalTeams` in my component code because `undefined` is easier to deal with. This isn't ideal, but if we can't use `undefined` or `null`, then I guess it isn't so bad. Forcing the developer to do this means they acknowledge what is going on and aren't being fooled into thinking their trpc procedure correctly returned `null` or `undefined`.

**Comment by dai-shi** [comment]
Hi, just added a commit in #14.
Now, it's `undefined` by default, which is configurable with options.

**Comment by asutula** [comment]
Yes, that is great! Tested it and seems to be working. Thank you.

**Comment by dai-shi** [comment]
Published: https://www.npmjs.com/package/jotai-trpc/v/0.5.1

**Comment by drMonro** [comment]
Can you, please, make working demo with what feature?
